### DISCUSSION TITLE
Create conditional return or throw

### DISCUSSION THREAD

**Discussion by geerttennapel** [dotnet/csharplang#7962] [Language Ideas]
## Summary

I know that this has been passed before, but the last active one was from 2019, so maybe times have changed?
I would like to make a nice compact syntax to return a value or throw an exception based upon a condition

## Motivation

Often developers are stuck writing if statement to make exits early on in a method by returning a value or throwing an error.
For example

```cs
public class MyClass
{
  public bool MyValidation(bool var1, bool var2, string someText)
  {
    if(string.IsNullOrEmpty())
    {
      throw new MyException("Missing argument `someText` required for validation");
    }

    if(!var1 && !var2)
    {
      return false;
    }

    if(var1 && var2) 
    {
      return true;
    }

    ... ...  (∩｀-´)⊃━☆ﾟ.*・｡ﾟ Magic
  }
}
```
The language already made this a bit better by onelining these situations
For example
```cs
public class MyClass
{
  public bool MyValidation(bool var1, bool var2, string someText)
  {
    if(string.IsNullOrEmpty()) throw new MyException("Missing argument `someText` required for validation");

    if(!var1 && !var2) return false;

    if(var1 && var2) return true;

    ...  (∩｀-´)⊃━☆ﾟ.*・｡ﾟ Magic

  }
}
```

## Detailed Design

BUT... what is we could do this

```cs
public class MyClass
{
  public bool MyValidation(bool var1, bool var2, string someText)
  {
    throw new MyException("Missing argument `someText` required for validation") when string.IsNullOrEmpty();

    return false when !var1 && !var2;

    return true when (var1 && var2);

    ...  (∩｀-´)⊃━☆ﾟ.*・｡ﾟ Magic

  }
}
```
this is even more intresting when we add pattern matching
for example

```cs
public class MyClass
{
  public bool MyValidation(bool var1, bool var2, string someText)
  {
    return GetSomeValue() when not null;

    return GetSomeOtherValue() when { SomeProperty = 10 };

    var val = GetSomeObject();
    return 20 when val is { SomeThing = true }


    ...  (∩｀-´)⊃━☆ﾟ.*・｡ﾟ MOAR MAGIC

  }
}
```

**Comment by sharpchen** [comment]
Use a switch expression instead
```cs
public class MyClass
{
  public bool MyValidation(bool var1, bool var2, string someText)
  {
    return (var1, var2, someText) switch 
    {
        (true, true, _) => true,
        (false, false, _) => false,
        (_, _, var text) when string.IsNullOrEmpty(text) => throw null!
        _ => throw null!
    }
  }
}
```

**Reply by geerttennapel** [reply]
you are answering a example. These methods can be way more complext. That why i added the magic

**Reply by iam3yal** [reply]
@geerttennapel And? how does this change anything?

**Comment by HaloFour** [comment]
See: https://github.com/dotnet/csharplang/discussions/138, https://github.com/dotnet/csharplang/discussions/325, https://github.com/dotnet/csharplang/issues/1368, https://github.com/dotnet/csharplang/issues/2966

**Reply by HaloFour** [reply]
This proposal doesn't address _why_ we'd want to add a feature like this to the language.

**Comment by huoyaoyuan** [comment]
I put strong down votes about changing the order of conditions like that. It's way too Englism or European-ism.

If the condition is tested first, it should appear at first. This is the nature order that fits execution order.
